torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+            ps_memory[p_idx : p_idx + n * matrix_approximation_rank].view(
+                n, matrix_approximation_rank
+            )
+        )
+        qs.append(
+            qs_memory[q_idx : q_idx + m * matrix_approximation_rank].view(
+                m, matrix_approximation_rank
+            )
+        )
+        p_idx += n * matrix_approximation_rank
+        q_idx += m * matrix_approximation_rank
+
+    # Initialize and then orthogonalize Qs.
+    with torch.random.fork_rng(devices=[]):
+        # Fork this RNG to avoid changing the seed globally and affecting the random sampling anywhere else in the training.
+        # The seed makes sure that the initial random values are the same across all the DDP replicas.
+        # Such seed should differ at every step.
+        # Since it is very slow to fork RNG state across all the CUDA devices,
+        # only fork on CPU and then move the generated tensor to the CUDA device.
+        torch.manual_seed(state.rng.randint(1_000_000_000))
+        for q in qs:
+            q.data = torch.randn(
+                *q.shape,
+                device="cpu",
+                dtype=dtype,
+            ).to(device)
+            _orthogonalize(q)
+
+    # Compute Ps.
+    for tensor, q, p in zip(high_rank_tensors, qs, ps):
+        torch.matmul(tensor, q, out=p)
+
+    # This allreduce is only applied to rank-1 tensors,
+    # so it should have been kicked off before the above computation on the high-rank tensors to hide more communication costs.
+    # However, this somehow requires a separate future chain at this time.
+    allreduce_contiguous_rank1_tensors_fut = dist.all_reduce(
+        rank1_tensors_memory, group=group_to_use, async_op=True
+    ).get_future()
+
+    def unpack_rank1_tensors_and_allreduce_ps(fut):
+        rank1_tensors_memory = fut.value()[0].div_(world_size)
+        idx = 0
+        for tensor in rank1_tensors:
+            tensor.copy_(rank1_tensors_memory[idx : idx + tensor.shape[0]])
+            idx += tensor.shape[0]
+
+        # Since these Ps will be orthogonized later, no need to divide them by world size.
+        return [
+            dist.all_reduce(ps_memory, group=group_to_use, async_op=True)
+            .get_future()
+            .wait()[0]
+        ]
+
+    def compute_qs(fut):
+        ps_memory = fut.wait()[0]
+        for p in ps:
+            _orthogonalize(p)
+
+        # Compute Qs.
+        for tensor, p, q in zip(high_rank_tensors, ps, qs):
+            torch.matmul(tensor.t(), p, out=q)
+
+        # Allreduce Qs.
+        return [
+            dist.all_reduce(qs_memory, group=group_to_use, async_op=True)
+            .get_future()
+            .wait()[0]
+        ]
+
+    def decompress(fut):
+        qs_memory = fut.wait()[0].div_(world_size)
+
+        for p, q, tensor in zip(ps, qs, high_rank_tensors):
+            torch.matmul(p, q.t(), out=tensor)
+            assert not torch.any(torch.isnan(tensor))
+        return [input_tensor]
```

Comment:
Why are we returning input_tensor here? Are we overwriting this tensor somewhere? Shouldn't we return the approximation PQ^T which is `tensor` in this case?

Reply:
Ah I see, every `tensor` in high_rank_tensors is a view into `input_tensor` and modifying tensor also modifies the appropriate per-param tensor in `input_tensor`. If this is the case, could we clarify that is happening here (and why it's preferable to do this instead of directly return the high_rank_tensors here?

Reply:
`high_rank_tensors` is a partial view of the input tensor (which is also the output here as the compression occurs in place). We need to return the input tensor, which is a combination of high-rank tensors and rank-1 tensors. This is the final output.